### PR TITLE
Cache the fact that sanity checks were performed for subprocesses

### PR DESCRIPTION
**Pull request by atrosinenko**
Half a year ago there was a discussion in #5042 on how to slightly speed up compilation and tests by caching the fact that the sanity checks were performed once by `emmake` / `emconfigure` / `tests/runner.py` and not rechecking on every `emcc` invocation. I have written a draft then but stumbled upon proper implementation for `test_sanity`.

Now I have slightly improved that implementation but is still unsure on whether it works properly for `sanity` test suite -- every invocation of it takes about 40-50 min on my machine not to say I don't know precisely what to pay attention too. The proposed implementation gives 4-8% speedup for some tests in my setup but definitely needs review. For example, what will be if we run non-WASM tests and then WASM ones and are `os.environ` manipulations thread-safe (or test runner spawns separate processes?).

### CODE REVIEW

**Comment by atrosinenko** [comment]
*pretending to be static analyzer:* [this](https://github.com/atrosinenko/emscripten/blob/c79e4ed20b66fd8b5d3fae3d2693dd69a8e7cceb/tools/shared.py#L562) line looks quite suspicious compared to [this](https://github.com/atrosinenko/emscripten/blob/c79e4ed20b66fd8b5d3fae3d2693dd69a8e7cceb/tools/shared.py#L558) one: on the left we check for existence of `PYTHON` and on the right the `cmd + '.exe'` is mentioned.

**Comment by kripken** [comment]
I think this is a very good idea!

There might be a simpler way to do it, but I'm not sure: `check_sanity` already looks at `EMCC_SKIP_SANITY_CHECK`. If it's set, we skip the check. So I think we can just set it after we check, and as a result, when we create child processes it will be set and they'll skip it?

**Comment by atrosinenko** [comment]
By using tags instead of an existing skip flag I tried to not skip checks
that are not identical to what was already checked, but I'm not sure my
implementation is either required or complete. Now I enumerated two types
of sanity check control flow: when WASM is either enabled or not - these
two cases are associated with two existing tags (not counting the special
ignore tag).

22 янв. 2018 г. 10:24 ПП пользователь "Alon Zakai" <notifications@github.com>
написал:

**Comment by kripken** [comment]
What do you mean by "a check if wasm is enabled or not"? (what part of the code are you referring to?)

**Comment by aheejin** [comment]
I guess he means he wanted to separately check the case when `~/.emscripten_cache/asmjs` has been checked vs. `~/.emscripten_cache/wasm` has. By the way, is the old `EMCC_SKIP_SANITY_CHECK` flag redundant then?

**Comment by atrosinenko** [comment]
I mean that technically [here](https://github.com/atrosinenko/emscripten/blob/c79e4ed20b66fd8b5d3fae3d2693dd69a8e7cceb/tools/shared.py#L515) and [here](https://github.com/atrosinenko/emscripten/blob/c79e4ed20b66fd8b5d3fae3d2693dd69a8e7cceb/tools/shared.py#L368), for example, the sanity check behavior depends on the value of `Settings.WASM_BACKEND`. But maybe it is too defensive (or, conversely, not complete).

I'm not sure the `EMCC_SKIP_SANITY_CHECK` is _completely_ redundant, because it allows to disable even the first sanity check. :) I'm not sure someone will ever need this, though. :)

**Comment by atrosinenko** [comment]
Hmm... `emmake` starts, performs checks and sets simple tag, then `emcc` is invoked with WASM backend, adds `tag_wasm` that disappears when `emcc` exits, then another `emcc` starts and performs its checks... I think, `emmake` most probably will set the simple tag regardless of what will be performed in child processes?

Maybe we should modify: `def check_sanity(..., fullCheck = False)`. When `fullCheck` is `False` we perform check as usual, when it is `True` (`emconfigure` / `emmake` / `tests/runner.py`) we check everything (all branches) and simply set `EMCC_SKIP_SANITY_CHECK=1` (we need to handle something like `EMCC_SKIP_SANITY_CHECK=-1` for `test_sanity`). But I fear, the check for wasm backend will most often fail and frighten users with warnings for mode that they will not ever use (until it stabilizes). If this will be implemented, we can probably drop [this check](https://github.com/kripken/emscripten/blob/incoming/tests/test_core.py#L7482).

**Comment by kripken** [comment]
`Settings.WASM_BACKEND` should be set very early, before the sanity checks - unless we have a bug. But reading through the code again now it looks correct to me. What specifically concerns you about the wasm backend part of this?

Overall I'd rather change this code as little as possible, as it's hard to test. We do have tests in `sanity`, but we don't run them constantly and coverage may be partial. And it seems like a minimal change to just set the env var after running the tests would let us not run the tests in child processes after the parent has.

**Comment by atrosinenko** [comment]
I mean that is `Settings.WASM_BACKEND` value controlled by emsdk_env.sh? If
not then emmake probably knows nothing about it when performing its checks
since `-s SETTING=value` options are passed to emcc but not emmake.

**Comment by kripken** [comment]
Do you mean https://github.com/juj/emsdk/blob/master/emsdk_env.sh ?

I'm not familiar with the emsdk code myself. But I don't think the emsdk needs to do anything with `Settings`, it just wraps around `emcc` etc. which do?

**Comment by saschanaz** [comment]
`Settings.WASM_BACKEND` is set in the initialization process of `Settings`:

https://github.com/kripken/emscripten/blob/c79e4ed20b66fd8b5d3fae3d2693dd69a8e7cceb/tools/shared.py#L1167-L1168

https://github.com/kripken/emscripten/blob/c79e4ed20b66fd8b5d3fae3d2693dd69a8e7cceb/tools/shared.py#L959-L961

https://github.com/kripken/emscripten/blob/c79e4ed20b66fd8b5d3fae3d2693dd69a8e7cceb/tools/shared.py#L950-L955

AFAIK it's effectively a static value determined when use install EMSDK tools, we shouldn't need to check both branches.

**Comment by juj** [comment]
This looks great since it is being observed to improve performance. Happy to see this kind of improvement to land. As a small sidenote though, I've been hoping that we could at some point delete the whole/most of the runtime sanity check code, apart from the version checks, it feels like we are adding code to do somewhat paranoid checks.

Rather, it might be nice to have a specific command `emcc --sanity_check` that did that, which would be something that developers could run to manually check their `node`, `java` etc. versions, something that would be easy to do if they run into cryptic failures?

**Comment by atrosinenko** [comment]
@saschanaz Thank you very much for your explanation. According to it and [this](https://github.com/kripken/emscripten/blob/0a6808161a68a65a53d003f803676fb47de7146e/emcc.py#L686) line setting `WASM_BACKEND` through `-s` option is explicitly prohibited, so assuming user will not set `EMCC_WASM_BACKEND` explicitly in their scripts executed **under** `emmake` / `emconfigure`, it is much more handy to do as @kripken suggested and just set `EMCC_SKIP_SANITY_CHECK=1` after the first check. Yes, there are some corner cases such as manually adjusting environment somewhere in between `emmake` and `emcc` on process tree, but hope they are not so significant.

**Comment by atrosinenko** [comment]
Simplified as suggested by @kripken and used `EMCC_SKIP_SANITY_CHECK=-1` (anything that is not '1', in fact) as indication that sanity check should not be cached. Under other conditions I set the `EMCC_SKIP_SANITY_CHECK` in the very beginning of `check_sanity`, so no need to account for every early return -- even if check fails just after this. This prevents console from being flooded (was it flooded previously? I don't know) with warnings when check fails by subsequent tool invocations (what is not very good, IMHO), but significantly simplifies the code. I have additionally specified `force=True` argument of `check_sanity` in some testing code. It is still seems to speed things up even with such arg, but it would be interesting if someone would test performance in real-world scenarios. I still invoke `check_sanity` without any args in `emmake` / `emconfigure`, etc. -- should it be changed to something like `force=DEBUG` as in `emcc.py`?

PS: Isn't [this line](https://github.com/atrosinenko/emscripten/blob/c79e4ed20b66fd8b5d3fae3d2693dd69a8e7cceb/tools/shared.py#L562) suspicious?

File: tests/runner.py [hunk 1]
```diff
@@ -213,6 +213,9 @@ def setUp(self):
         if temp_file.endswith('.ll'):
           self.has_prev_ll = True
 
+    # Cache check results
```
Comment (kripken) [inline tests/runner.py]: Doing this for each test is more than we need, I think. We can just do it once at the very beginning, at the top level?

File: tests/test_other.py [hunk 2]
```diff
@@ -7926,6 +7926,11 @@ def test_wasm_backend(self):
     if old == '1': return # already the default
     try:
       os.environ['EMCC_WASM_BACKEND'] = '1'
+      if EMCC_SKIP_SANITY_CHECK_NAME in os.environ:
+        # Cached sanity check result was for different configuration
+        del os.environ[EMCC_SKIP_SANITY_CHECK_NAME]
```
Comment (kripken) [inline tests/test_other.py]: it's usually better to skip the test with a message than to modify the environment.

in any case, though, I don't think we need to check sanity here?

File: tests/test_sanity.py [hunk 3]
```diff
@@ -29,13 +29,17 @@ def setUpClass(self):
     print('WARNING: This will modify %s, and in theory can break it although it should be restored properly. A backup will be saved in %s_backup' % (EM_CONFIG, EM_CONFIG))
     print()
 
+    os.environ[EMCC_SKIP_SANITY_CHECK_NAME] = '-1'
```
Comment (kripken) [inline tests/test_sanity.py]: -1?

File: tools/shared.py [hunk 4]
```diff
@@ -1497,6 +1502,9 @@ def has_substr(array, substr):
 
   @staticmethod
   def configure(args, stdout=None, stderr=None, env=None):
+    # Check once and cache results
+    check_sanity()
```
Comment (kripken) [inline tools/shared.py]: these also seem unnecessary?

File: tests/runner.py (hunk 1 above)
Comment (atrosinenko) [reply tests/runner.py]: Oops, I probably put it in the wrong place. I thought to place it to the global setup function.

File: tests/test_other.py (hunk 2 above)
Comment (atrosinenko) [reply tests/test_other.py]: As I understand, we already skip this test if no environment modification (that was in place before this PR) is required. :) As you already noted, we don't need to perform checks before every test case. I redo sanity check **for this particular test case** because here we **artificially changed** the condition on which sanity check control flow depends (in fact, I don't handle very well another input: the `force` function argument here and in the other places).

File: tests/test_sanity.py (hunk 3 above)
Comment (atrosinenko) [reply tests/test_sanity.py]: I agree, this `-1` looks just like a typo, but in fact, it (and every other string) has opposite meaning to `1`. It would be better to replace it with `os.environ[EMCC_SKIP_SANITY_CHECK_NAME] = 'dontskip'`

File: tools/shared.py (hunk 4 above)
Comment (atrosinenko) [reply tools/shared.py]: Hmm, according to comments, `emconfigure` (this `configure` function is its main function, yes?) performs builds with regular `clang`... Maybe it is really unnecessary here. On the other hand, would it take significant amount of time to check sanity once in case some day `emcc` will eventually be used instead of `clang` (meanwhile, is native `clang` invoked directly or proxyed by `emcc.py` now?).

**Comment by kripken** [comment]
Thinking some more here, maybe I wasn't clear enough before. What I was proposing we do is this: https://github.com/kripken/emscripten/commit/dfec072ede233280c8932bfc5f0144af37319749

That appears to achieve the goal here, for example when building a system lib I see that the sanity checks are no longer done after the first one. Is there something that commit doesn't do that this PR does?

**Comment by atrosinenko** [comment]
Why my PR is more complicated:
* dfec072 sets `EMCC_SKIP_SANITY_CHECK` unconditionally, so most probably alters the behavior of `sanity` test suite (but maybe it is OK)
* it doesn't account for that compiler in `test_wasm_backend` wants to perform _different_ checks comparing to what have been done and cached because of environment changes
* it does not run checks explicitly in `runner.py`/`emmake`/`emconfigure`, so are they really cached somewhere? (but maybe they are anyways performed somewhere there)

I realize that this may be overcomplicated (especially the first one) because I'm not very familiar with Emscripten internals. On the other hand, failing to cache the check in `emmake`, for example, will probably result in some (not very large, though) performance regression that no one will know about, so it may be worth to cache sanity as explicitly as possible. :) But simplifications are welcome, of course. :)

**Comment by kripken** [comment]
Thanks, now I understand you a lot better I think.

 * Good point about the sanity and `other.test_wasm_backend` tests, yeah, some of those do need to run sanity checks in child processes. I think they can just unset that env var, as you did in one place, yeah. Running the tests, the relevant ones that fail with that commit are `sanity.test_emcc sanity.test_emcc_caching`.
 * I don't see why runner.py, emmake and emconfigure are special? I don't think we need to do anything for them in this PR.
 * I think it's ok to set `EMCC_SKIP_SANITY_CHECK` unconditionally if they run at least once, as in dfec072 - that should be safe. If some sanity checks need to be updated, we may need to do that, but it's better to do more work to update them and keep the main code simpler, I think.

**Comment by atrosinenko** [comment]
> Good point about the sanity and other.test_wasm_backend tests, yeah, some of those do need to run sanity checks in child processes. I think they can just unset that env var, as you did in one place, yeah.

> I think it's ok to set EMCC_SKIP_SANITY_CHECK unconditionally if they run at least once

With this behavior of setting `EMCC_SKIP_SANITY_CHECK` only if it is undefined I tried to replicate the old behavior of `sanity` test suite in an error-proof way while knowing close to nothing about this test suite. :) This logic is intended to conditionally replicate no-cache behavior in case when before PR the checks were performed **multiple** times. Did I get you right: you suggest to update the `sanity` test suite according to new code. If yes, then I agree, it is probably much more correct way to solve this issue.

> I don't see why runner.py, emmake and emconfigure are special? I don't think we need to do anything for them in this PR.

I'm not sure these scripts do not check sanity on their own already, but if not, then, let's say `emmake` runs some script that invokes `emcc` 10 times:
```
emmake
|-- emcc 1
|-- emcc 2
...
|-- emcc 10
```
As far as I understand the environment variables semantics, `emcc 1` cannot influence `emcc 2`, `emcc 3`, etc., but all of them inherit the environment of `emmake`, so it is `emmake` that should set `EMCC_SKIP_SANITY_CHECK=1` (excuse me if I don't understood right your question).

**Comment by kripken** [comment]
> Did I get you right: you suggest to update the sanity test suite according to new code. If yes, then I agree, it is probably much more correct way to solve this issue.

Yes, exactly. (Unless we are missing something, it's possible that when updating the tests we'll see some corner case that is checked there.)

> As far as I understand the environment variables semantics, emcc 1 cannot influence emcc 2, emcc 3, etc., but all of them inherit the environment of emmake, so it is emmake that should set `EMCC_SKIP_SANITY_CHECK=1` (excuse me if I don't understood right your question).

Yes, I think you are right about this example and how these things work. This also helped me understand you better: I think your goal is higher than mine, I didn't even consider optimizing the emmake/emconfigure case yet :) I think it's reasonable to do so, but how about leaving it for followup work? I believe it is safer to do this in small pieces. So I suggest as a first stage we do just the simple setting of the skip env var after checking sanity, that will already help a lot by itself.

**Comment by atrosinenko** [comment]
> I think your goal is higher than mine, I didn't even consider optimizing the emmake/emconfigure case yet :) I think it's reasonable to do so, but how about leaving it for followup work?

Do you mean revert these two-line fixes for `emmake` / `emconfigure` / `embuilder` in this PR, merge the main change and send these adjustments in the following PR? If so, then what is the scope of this PR -- faster running tests, yes?

**Comment by kripken** [comment]
Yes, I think faster running tests is a good goal for this first PR. And we can leave emmake/etc. to later, that will be more complex.

**Comment by aheejin** [comment]
Are we doing this caching of sanity check today? It's an old issue and things might have changed, so I'm wondering.

**Comment by sbc100** [comment]
Looks like we do check in each sub-process..  so lets keep this open
